BUILD-2909-hudson.model.ParametersDefinitionProperty

NatWoodson opened the pull request:
## Ticket

[JIRA-2909](https://jira.tinyspeck.com/browse/BUILD-2909)

## Overview

Noticed that nothing under the  hudson.model.ParametersDefinitionProperty section of the xml was being translated.

<img width="411" alt="Screen Shot 2022-10-20 at 1 40 41 PM" src="https://user-images.githubusercontent.com/112515811/197037046-5b67a6c8-878f-41a0-bb6e-0a08eaf294b9.png">

 The `<pre>` tag was added within a "description" block in the UI for formatting so it's now added a parameter in order to convert correctly.

## Testing

Test XML Used:
```
<project>
    <actions/>
    <description>Test</description>
    <keepDependencies>false</keepDependencies>
    <properties>
        <hudson.model.ParametersDefinitionProperty>
            <parameterDefinitions>
                <hudson.model.ChoiceParameterDefinition>
                    <name>DEPLOY_GROUP</name>
                    <description><pre>Group of hosts to upgrade: Testing</pre></description>
                    <choices class="java.util.Arrays$ArrayList">
                        <a class="string-array">
                            <string>role:notraffic</string>
                            <string>dogfood:all</string>
                            <string>role:pepper</string>
                            <string>role:heavy_pepper</string>
                            <string>smart</string>
                        </a>
                    </choices>
                </hudson.model.ChoiceParameterDefinition>
                <hudson.model.BooleanParameterDefinition>
                    <name>DRY_RUN</name>
                    <description>Make this a dry run that shows the list of hosts that would be upgraded</description>
                    <defaultValue>true</defaultValue>
                </hudson.model.BooleanParameterDefinition>
                <hudson.model.StringParameterDefinition>
                    <name>DEPLOY_HOSTS</name>
                    <description>Deploy to specified hosts</description>
                    <trim>false</trim>
                </hudson.model.StringParameterDefinition>
            </parameterDefinitions>
        </hudson.model.ParametersDefinitionProperty>
    </properties>
</project>
```

DSL Output:
```
job("Test") {
	description("Test")
	keepDependencies(false)
	properties {
		parameters {
			parameterDefinitions {
				choiceParam {
					name("DEPLOY_GROUP")
					description("Group of hosts to upgrade: Testing")
					choices(["role:notraffic", "dogfood:all", "role:pepper", "role:heavy_pepper", "smart"])
				}
				booleanParam {
					name("DRY_RUN")
					description("Make this a dry run that shows the list of hosts that would be upgraded")
					defaultValue(true)
				}
				stringParam {
					name("DEPLOY_HOSTS")
					description("Deploy to specified hosts")
					trim(false)
				}
			}
		}
	}
}


No untranslated tag! Fit for moving
```

Commit:
add gitLFSPull attribute

Commit:
edit attributes related to parameterDefinitions

Commit:
add pre attribute as parameter type

Commit:
remove trim property from skipped list and delete property being worked on other branch

pamgluss-slack commented:
Can you update the `DSL Output` section of your PR description? I don't see `pre` there so i'm not sure what it should look like

pamgluss-slack commented on src/main/resources/translator.properties:
```diff
@@ -239,19 +239,21 @@ credentialType = type
+pre.type = PARAMETER
+
 hudson.model.BooleanParameterDefinition = booleanParam
-hudson.model.BooleanParameterDefinition.type = com.adq.jenkins.xmljobtodsl.dsl.strategies.custom.DSLParamStrategy
```
Same question from your previous PR -- why were we using the custom type before? Was it doing anything we need to pay attention to?

NatWoodson replied:
In the [Jenkins api viewer](https://jenkins.tinyspeck.com/plugin/job-dsl/api-viewer/index.html#method/javaposse.jobdsl.dsl.helpers.BuildParametersContext.choiceParam) the properties for choiceParam, booleanParam and stringParam are formatted as parameters so that's what the custom class is meant to do. 

<img width="687" alt="Screen Shot 2022-10-25 at 4 49 18 PM" src="https://user-images.githubusercontent.com/112515811/197879056-79ccfcab-3840-4cc6-b59d-3de0d0521c92.png">
<img width="710" alt="Screen Shot 2022-10-25 at 4 50 00 PM" src="https://user-images.githubusercontent.com/112515811/197879181-d70d9a17-7dcb-4f04-a410-5556958b233f.png">

In the [Tinyspeck api viewer](https://jenkins.tinyspeck.com/plugin/job-dsl/api-viewer/index.html#method/javaposse.jobdsl.dsl.jobs.FreeStyleJob.parameters) they are objects. 

<img width="370" alt="Screen Shot 2022-10-25 at 4 51 53 PM" src="https://user-images.githubusercontent.com/112515811/197879530-e14b8b7a-79bb-4d8c-acb9-e15aad1a0476.png">

Even with the custom classes in place, nothing is displaying when the project is run. I'll investigate further to try and find out exactly what is causing that bug.

But looking at these differences between the TS api and the Jenkins should be stick with the formatting on the TS site?

NatWoodson commented:
> Can you update the `DSL Output` section of your PR description? I don't see `pre` there so i'm not sure what it should look like

The `pre` tag is an html tag that was added in the UI of Jenkins to format the text within a "description" block so there is no equivalent in the DSL. I linked this visual in the ticket but I'll leave it here as well. Because it's an html formatting tag, I added it as a parameter so that the description would still show up, but it's not formatted. 

<img width="411" alt="Screen Shot 2022-10-20 at 2 56 22 PM" src="https://user-images.githubusercontent.com/112515811/197874430-59b3e4ea-8ec8-4406-974d-bc821a7447c9.png">

pamgluss-slack commented:
I see. Then `<pre>` shouldn't be considered a DSL tag at all, I think we should handle it as a unique case, like `<b>` or `<i>` https://www.w3schools.com/tags/tag_b.asp

NatWoodson commented:
> I see. Then `\<pre\>` shouldn't be considered a DSL tag at all, I think we should handle it as a unique case, like `<b>` or `<i>` https://www.w3schools.com/tags/tag_b.asp

Yes, so as of now I added it as a parameter because the description was completely getting skipped over. Should we plan on removing that from translator.properties and handling these tags a different way during conversion?

pamgluss-slack commented:
> 

Yeah it's not really an XML tag at all, it's an HTML tag that's accidentally being included. We should skip it in translator.properties and put a ticket on the backlog to keep in included w/ the description

Commit:
add pre to ignored and remove from translator properties

NatWoodson commented:
Ok, I removed `pre` from translator.properties and added it to ignored, but the text within description is not rendering because it's within that `pre` tag so I'm going to look into how to fix this now without having it in the translator.properties 
```
parameterDefinitions {
				choiceParam {
					name("DEPLOY_GROUP")
					description()
					choices(["role:notraffic", "dogfood:all", "role:pepper", "role:heavy_pepper", "smart"])
				}
				booleanParam {
					name("DRY_RUN")
					description("Make this a dry run that shows the list of hosts that would be upgraded")
					defaultValue(true)
				}
				stringParam {
					name("DEPLOY_HOSTS")
					description("Deploy to specified hosts")
					trim(false)
				}
```

Commit:
removing pre from ignored and adding back to translator.properties

NatWoodson commented:
I added `pre` back to translator.properties as a parameter so that the description displays correctly until we deal with the custom class for handling HTML.
